src/msg_recorder/recorder.cc

```diff
@@ -10,25 +10,88 @@
 
 namespace cris::core {
 
-MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner)
+MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, const int& snapshot_interval, std::shared_ptr<JobRunner> runner)
     : Base(std::move(runner))
-    , record_dir_(record_dir_prefix / RecordDirNameGenerator()) {
+    , record_dir_(record_dir_prefix / RecordDirNameGenerator())
+    , record_strand_(runner ? runner->MakeStrand() : MakeStrand())
+    , snapshot_interval_(snapshot_interval)
+    , snapshot_thread_(std::bind(&MessageRecorder::SnapshotStart, this)) {
     std::filesystem::create_directories(record_dir_);
 }
 
+void MessageRecorder::SnapshotStart() {
+    while (!snapshot_shutdown_flag_.load()) {
+        AddJobToRunner([this]() { GenerateSnapshot(keep_max_); }, record_strand_);
+        std::this_thread::sleep_for(std::chrono::seconds(snapshot_interval_));
+    }
+}
+
+void MessageRecorder::SnapshotEnd() {
+    snapshot_shutdown_flag_.store(true);
+    if (snapshot_thread_.joinable()) {
+        snapshot_thread_.join();
+    }
+    record_init_datas.clear();
+}
+
 MessageRecorder::~MessageRecorder() {
+    SnapshotEnd();
     files_.clear();
-
     if (std::filesystem::is_empty(GetRecordDir())) {
         LOG(INFO) << "Record dir " << GetRecordDir() << " is empty, removing...";
         std::filesystem::remove(GetRecordDir());
     }
 }
 
+void MessageRecorder::GenerateSnapshot(const int& max) {
+    // remove symlinks and close DB
+    for (const auto& file : files_) {
+        if (!file->GetFilePath().empty() && !std::filesystem::is_empty(file->GetFilePath())) {
+            std::filesystem::remove_all(file->GetFilePath());
+        }
+    }
+    files_.clear();
+
+    // create dir for individual interval
+    std::filesystem::path interval_folder = record_dir_ / std::string("Snapshot") / std::string("HOURLY");
+    if (!std::filesystem::exists(interval_folder)) {
+        std::filesystem::create_directories(interval_folder);
+    }
+
+    std::filesystem::directory_entry withdraw;
+    int                              counter = 0;
+    for (auto const& dir_entry : std::filesystem::directory_iterator{interval_folder}) {
+        counter++;
+        if (!withdraw.exists() ||
+            (withdraw.last_write_time().time_since_epoch().count() -
+             dir_entry.last_write_time().time_since_epoch().count()) > 0) {
+            withdraw = dir_entry;
+        }
+    }
+
+    // withdraw the earliest snapshot if needed
+    if (counter >= max) {
+        std::filesystem::remove_all(withdraw);
+    }
+
+    const auto snapshot_dir = interval_folder / SnapshotDirNameGenerator();
+    std::filesystem::create_directories(snapshot_dir);
+    std::filesystem::copy(record_dir_, snapshot_dir, std::filesystem::copy_options::recursive | std::filesystem::copy_options::create_symlinks);
```

Comment:
Compiler gives:
```
  what():  filesystem error: cannot copy: Is a directory [/tmp/CRTestTmpDir.12/record.20221121-014339.UTC.pid.12] [/tmp/CRTestTmpDir.12/record.20221121-014339.UTC.pid.12/Snapshot/HOURLY/20221121-014340.UTC]
```
 when I tried to use symlinks as the copy option, otherwise the dir name is too long for filesystem to copy.

Reply:
copy symlink，不是createsymlink。https://en.cppreference.com/w/cpp/filesystem/copy_options

create_symlinks: Instead of creating copies of files, create symlinks pointing to the originals. Note: the source path must be an absolute path unless the destination path is in the current directory.

copy_symlinks:	Copy symlinks as symlinks, not as the files they point to

Reply:
Changes to `copy_symlinks`, back to the name-too-long error:
```
  what():  filesystem error: cannot copy: File name too long [/tmp/CRTestTmpDir.12/record.20221121-025941.UTC.pid.12] [/tmp/CRTestTmpDir.12/record.20221121-025941.UTC.pid.12/Snapshot/HOURLY/20221121-025941.UTC]

```

Reply:
@xkszltl Any ideas?